src/Expect.elm

```diff
@@ -546,84 +530,90 @@ reportFailure comparison expected actual =
         |> String.join "\n"
 
 
-reportCollectionFailure : String -> a -> b -> List c -> List d -> String
-reportCollectionFailure comparison expected actual missingKeys extraKeys =
-    [ toString actual
-    , "diff:" ++ formatDiffs Missing missingKeys ++ formatDiffs Extra extraKeys
-    , "╷"
-    , "│ " ++ comparison
-    , "╵"
-    , "diff:" ++ formatDiffs Extra missingKeys ++ formatDiffs Missing extraKeys
-    , toString expected
-    ]
-        |> String.join "\n"
-
-
-type Diff
-    = Extra
-    | Missing
-
+getFailureMessage : { given : String, description : String, reason : Test.Expectation.Reason } -> String
+getFailureMessage { given, description, reason } =
+    case reason of
+        Test.Expectation.Custom ->
+            description
+
+        Test.Expectation.Equals e a ->
+            verticalBar description e a
+
+        Test.Expectation.Comparison e a ->
+            verticalBar description e a
+
+        Test.Expectation.ListDiff e a ( i, itemE, itemA ) ->
+            verticalBar description e a
+                ++ "\n\nThe first diff is at index index "
+                ++ toString i
+                ++ ": it was `"
+                ++ itemA
+                ++ "`, but `"
+                ++ itemE
+                ++ "` was expected."
+
+        Test.Expectation.CollectionDiff { expected, actual, extra, missing } ->
+            let
+                extraStr =
+                    if List.isEmpty extra then
+                        ""
+                    else
+                        "\nThese keys are extra: "
+                            ++ (extra |> String.join ", " |> \d -> "[ " ++ d ++ " ]")
+
+                missingStr =
+                    if List.isEmpty missing then
+                        ""
+                    else
+                        "\nThese keys are missing: "
+                            ++ (missing |> String.join ", " |> \d -> "[ " ++ d ++ " ]")
+            in
+                verticalBar description expected actual
+                    ++ "\n"
+                    ++ extraStr
+                    ++ missingStr
 
-formatDiffs : Diff -> List a -> String
-formatDiffs diffType diffs =
-    if List.isEmpty diffs then
-        ""
-    else
-        let
-            modifier =
-                case diffType of
-                    Extra ->
-                        "+"
 
-                    Missing ->
-                        "-"
-        in
-            diffs
-                |> List.map toString
-                |> String.join ", "
-                |> (\d -> " " ++ modifier ++ "[ " ++ d ++ " ]")
+{-| If the given expectation fails, replace its failure message with a custom one.
 
+    "something"
+        |> Expect.equal "something else"
+        |> Expect.onFail "thought those two strings would be the same"
+-}
+onFail : String -> Expectation -> Expectation
+onFail str expectation =
```

Comment:
🎨  `onFail` is better named `formatFailure`, inline with Result, or better yet, `mapError`